Copter: add 6DoF support

IamPete1 opened the pull request:
This adds support for copters that can fly in any direction and at any orientation.

https://youtu.be/CXIlZ8mCfGM

This adds a new attitude control and motor back-end and scripting bindings for each. Scripting can be used to set the true vehicle attitude while the existing navigation 'attitude' is used for the thrust vector. Scripting is also used to set the 6DoF matrix for the mixer. I did spend a some time trying to calculate the mixer directly from motor positions and trust vectors, but it all got too complex. The ideal mixer can be found and then loaded with the script. For my demo I used a PSO in MATLAB to find the mix matrix, the mix is not included.

Fixes #10117

IamPete1 commented:
For Copter + boosters rather than a dedicated 6DoF vehicle we might want some balance at the attitude control level, so it will let the vehicle lean a little in some cases or if it can only boost forwards but not backwards, I suspect all the edge cases there will be at least a complex as this support for dedicated 6DoF vehicles.

IamPete1 commented:
rebased, rates now use same offsets as angles. Added options to switch off lateral and forward control, so can now go 4DoF like a normal copter, 5DoF and 6DoF. Fixed `get_frame_mav_type()` so the correct number of motors for the new method, this fixes MP motor test which used the `MAV_TYPE` to determine the number of motors.

tridge commented:
very cool! maybe use HAL_MINIMIZE_FEATURES?

IamPete1 commented:
rebased and included https://github.com/ArduPilot/ardupilot/pull/15717 and https://github.com/ArduPilot/ardupilot/pull/16204 also put new mixer and attitude control in `ENABLE_SCRIPTING` ifdefs the mixer requires scripting and that mixer is the only user of the new 6DoF lib.

IamPete1 commented:
some more flying on this PR https://youtu.be/0p9jmrf1eFM

agentrfr commented:
Wouldnt it make more sense to have two contra rotating motors on each hardpoint instead of a single motor and relying on reversing prop direction? You could leave the motor/ESC at min throttle that is within the BLDC torque curve and avoid the giggling issue altogether

auturgy commented:
this might interest the Sub guys too: @jaxxzer

IamPete1 commented:
rebased, now includes mixer

IamPete1 commented:
changed control priority to be in earth frame, previously it was applying body frame throttle first, but this might not be up. Also fixed a rotation issue for combined roll and pitch offsets.

1ozzyabroad commented:
Great concept! Could it be adapted to use a servo controlling a variable pitch prop' to eliminate the "giggles'?

IamPete1 commented:
> Wouldnt it make more sense to have two contra rotating motors on each hardpoint instead of a single motor and relying on reversing prop direction? You could leave the motor/ESC at min throttle that is within the BLDC torque curve and avoid the giggling issue altogether

Then you have to carry 6 extra motors and esc's. But yes that would work.

> Great concept! Could it be adapted to use a servo controlling a variable pitch prop' to eliminate the "giggles'?

yes

tridge commented on ArduCopter/system.cpp:
```diff
@@ -556,8 +564,15 @@ void Copter::allocate_motors(void)
+    } else {
+        attitude_control = new AC_AttitudeControl_Multi(*ahrs_view, aparm, *motors, scheduler.get_loop_period_s());
+        ac_var_info = AC_AttitudeControl_Multi::var_info;
+    }
```
should change the panics to AP_BoardConfig::config_error() below

IamPete1 replied:
done

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:
```diff
@@ -0,0 +1,95 @@
+        _singleton = this;
+    }
+
+    virtual ~AC_AttitudeControl_Multi_6DoF() {}
```
we don't normally add a destructor.

IamPete1 replied:
removed

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp:
```diff
@@ -0,0 +1,327 @@
+
+void AP_MotorsMatrix_6DoF_Scripting::output_to_motors()
+{
+    int8_t i;
```
you can probably make this a uint8_t and move it down into the for loop line

IamPete1 replied:
done

tridge commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:
```diff
@@ -0,0 +1,95 @@
+
+    // Command a Quaternion attitude with feedforward and smoothing
+    // not used anywhere in current code, panic so this implementaiton is not overlooked
+    void input_quaternion(Quaternion attitude_desired_quat) override {
```
better to remove input_quaternion as not used

IamPete1 replied:
removed function

rmackay9 commented:
Looks generally OK.  I'll go over this in more detail over the next week.

FYI @lthall

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:
```diff
@@ -0,0 +1,87 @@
+    */
+
+    // Command an euler roll and pitch angle and an euler yaw rate with angular velocity feedforward and smoothing
+    void input_euler_angle_roll_pitch_euler_rate_yaw(float euler_roll_angle_cd, float euler_pitch_angle_cd, float euler_yaw_rate_cds)  override ;
```
tiny white space issue at the end of this line..

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.h:
```diff
@@ -0,0 +1,64 @@
+    // output_to_motors - sends minimum values out to the motors
+    void output_to_motors() override;
+
+    void set_roll_pitch(float roll_deg, float pitch_deg) override;
```
let's put a comment above the set_roll_pitch() method

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.h:
```diff
@@ -0,0 +1,64 @@
+    float _right_factor[AP_MOTORS_MAX_NUM_MOTORS];        // each motors contribution to right thrust
+
+    // true if motor is 3D, it can go from -Spin max to +Spin max, if false motor is can go from Spin min to Spin max
+    bool  _motor_3D[AP_MOTORS_MAX_NUM_MOTORS];
```
it's just naming but I guess this is a motor that uses a "3D" propeller so it can provide thrust when reversed?  Could we use a name that is more specific in terms of what it's unusual ability is.  something like "_can_reverse_thrust[]" or "_reversible[]" or something like this.  Sorry, I don't want to start you off on a guessing game of what I think is an acceptable name but when I see, "_motor_3D" I think that perhaps this is a 3D position of a motor or a 3D model of a motor.  It doesn't scream out to me that it is a motor that can provide reverse thrust.

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.h:
```diff
@@ -0,0 +1,64 @@
+    // store last values to allow deadzone
+    float _last_thrust_out[AP_MOTORS_MAX_NUM_MOTORS];
+
+    // Current offset angles
```
I guess these offsets are in degrees?  Let's make the units clear in either the variable name or the comments above the declaration.

IamPete1 replied:
there in radians so I did not bother with the units.

rmackay9 commented on libraries/AP_Motors/AP_Motors_Class.h:
```diff
@@ -98,6 +99,7 @@ class AP_Motors {
     void                set_throttle_filter_cutoff(float filt_hz) { _throttle_filter.set_cutoff_frequency(filt_hz); }
     void                set_forward(float forward_in) { _forward_in = forward_in; }; // range -1 ~ +1
     void                set_lateral(float lateral_in) { _lateral_in = lateral_in; };     // range -1 ~ +1
+    virtual void        set_roll_pitch(float roll_deg, float pitch_deg) {};
```
I think it will be easy for future devs to get confused between the use of "set_roll_pitch" and the individual "set_roll" and "set_pitch" calls so let's try and unify them or clarify the difference in the comments at least.

rmackay9 commented on libraries/AP_Scripting/examples/6DoF_roll_pitch.lua:
```diff
@@ -0,0 +1,54 @@
+-- This script sets the target roll and pitch to the current value when arming or when E-stop is removed
```
We should expand these comments a lot more maybe even provide a link to a wiki page where the 6DOF copter setup is explained.

I think ideally users should be able to look at this description and understand how the script should be used and its purpose.  So for example, a normal user might look at this and think it is useful in some way for a regular multicopter.

rmackay9 commented on libraries/AP_Scripting/examples/6DoF_roll_pitch.lua:
```diff
@@ -0,0 +1,54 @@
+function update() -- this is the loop which periodically runs
+
+  -- motor state
+  local current_active = false
```
this is a script so we are not as strict as with the c++ code but maybe rename "current_active" to "motors_armed" or "motors_spinning"?  Sorry to be so annoying about names but it saves the reader time if we pick good names.

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:
```diff
@@ -0,0 +1,87 @@
+    float get_althold_lean_angle_max() const override { return 9000.0f; }
+
+    // set the attitude that will be used in 6DoF flight
+    void set_offset_roll_pitch(float roll_deg, float pitch_deg) {
```
the comments are good above this method but the name of the method could be better I think.  These really are target lean angles I think so I'm not sure the "offset" part of the name is clear.  I guess the regular lean angle inputs that go into the overriden methods above are actually accelerations..

Maybe something like set_6d0f_roll_pitch_target().

IamPete1 replied:
They are offsets if your using 4DoF control, you can for example fly the frame like a normal copter where it rolls and pitches with stick inputs except the vehicle is at a arbitrary orientation.

rmackay9 commented on libraries/AC_AttitudeControl/AC_AttitudeControl_Multi_6DoF.h:
```diff
@@ -0,0 +1,87 @@
+        pitch_offset_deg = pitch_deg;
+    }
+
+    void set_forward_enable(bool b) {
```
It would be good to have a comment above "set_forward_enable()" and "set_lateral_enable()"

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp:
```diff
@@ -0,0 +1,325 @@
+
+
+    /*
+        upwards thrust, independent of orentaiton
```
small typo, "orentaiton" should be "orientation".

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp:
```diff
@@ -0,0 +1,325 @@
+
+
+    /*
+        rotations
```
when we say, "rotations" I guess this is the part of the motor mixer that causes the vehicle to rotate in roll, pitch yaw?

IamPete1 replied:
exactly, I will improve the comment

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp:
```diff
@@ -0,0 +1,325 @@
+
+    // set limit flags if output is being scaled
+    if (rpy_ratio < 1) {
+        limit.roll = 1;
```
normally we use "true" and "false" although technically they are actually uint8_t bits.. but let's use true/false anyway.

BTW, are these limits ever unset?

rmackay9 commented on libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp:
```diff
@@ -0,0 +1,325 @@
+
+}
+
+void AP_MotorsMatrix_6DoF_Scripting::set_roll_pitch(float roll_deg, float pitch_deg)
```
let's put a comment above this method implementation ..

rmackay9 commented:
It's an amazing vehicle and it is really great to see this PR.  I've put in some comments which are really all just about naming and commenting in the hopes that that makes it easier for other developers to understand what's going on.

The elephant in the room is the mixup between target roll and pitch angles VS forward and lateral accelerations.  I hope @lthall will provide some comments about this and whether it's acceptable to do what we've done in this PR or if we need to handle it more strategically.

IamPete1 commented:
@rmackay9 thanks to the review

re-target roll and pitch VS acceleration, the higher level code cannot tell because we set the targets based on the desired angle and then multiply back by throttle [here ](https://github.com/IamPete1/ardupilot/blob/a45cef17d109c651a1701bc4d7cd2b1a0195806c/libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp#L92). That does arbitrarily limit the vehicle but allows the higher level controllers to still apply. If we were un-couple lateral acceleration from vertical throttle so we could go full speed sideways with no vertical thrust, for example, we would then need a new PID to go from lateral acceleration to lateral thrust (Like the Z accel controller does for vertical thrust).

lthall commented:
I have been planning for this class of aircraft for a long time now and what you are proposing is exactly what I was thinking we should implement. How we realise it is less clear but what I was thinking was a vehicle specific transformation from target acceleration to a 6dof attitude controller.

For multicopter that would be a roll, pitch, thrust command where we can do what we want with yaw.

For this aircraft it would be a transform of the earth frame acceleration vectors to a body frame thrust command where we can do what we want with roll, pitch, and yaw.

The question is how to best implement this.....
I was thinking a generic 6dof attitude controller that has not only Z thrust but also X and Y.
Then in the position controller the transformation from acceleration to actuators would be done based on the frame. This may mean that we move the acceleration controller for alt hold into the attitude controller. This approach is also supported by the need to run the throttle pid loop at the same rate as the rate controllers. When we did the onion rewrite I was not sure if the acceleration PID should be in the position controller or the attitude controller, I think I put it in the wrong place...

IamPete1 commented:
Fixed all review issues and re-tested in realflight.

tridge commented:
@lthall is happy for input_quaternion to be removed on the understanding that @IamPete1 will help implement it for the 6DoF backend when it is implemented in the mavlink quaternion control

rmackay9 commented:
Merged, thanks!

huangfuluya commented:
I am making such a copter recently, but I wonder how these values are calculated in the Motors_6DoF.lua
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Scripting/examples/Motors_6DoF.lua
Can someone give an example about how to calculate this matrix for a new copter?

IamPete1 commented:
@huangfuluya Have you seen the conference talk, https://youtu.be/QUDhnYvH66k ? If you start a thread over on the forum I can give you some pointers.

huangfuluya commented:
Thanks，I have watched it many times, but I can't understand it very well. Maybe I haven't described my problem clearly. The matrix I calculated using my understanding is slightly different from the matrix in the demo, I want to know why these differences occur.  You mentioned here that you used a PSO in MATLAB to find the mix matrix. Is there any paper or programs about this matrix calculation for reference?

IamPete1 commented:
@huangfuluya start a thread over on the forum, if you can explain how your calculations are done I can see if I can spot any issue. Or I could run your vehicle through my calculations. https://discuss.ardupilot.org/

ryrobotics commented:
Hi Peter! What a great job! I am very interested in this kind of 6 DoF drone, and it can be a great tool for performing inspections. But I am not sure if it is possible to assemble extra visual sensors and onboard computers, like Xavier NX, on this frame. So could you share some drawings/detailed pictures of your drone or give me some hints? Thanks!
